### PR TITLE
Download agent

### PR DESCRIPTION
**Pull request by martinmilev**
Ready for review.

### CODE REVIEW

**Comment by freekman** [comment]
Is this PR ready for review

File: src/com/clouway/downloadAgent/DownloadAgent.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URL;
+import java.net.URLConnection;
+
+/**
+ * @Author Martin Milev (martinmariusmilev@gmail.com)
+ * @Date 11.07.16.
```
Comment (freekman) [inline src/com/clouway/downloadAgent/DownloadAgent.java]: remove the date

File: src/com/clouway/downloadAgent/DownloadAgent.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ */
+public class DownloadAgent {
+  private ProgressBar progressBar = new ProgressBar() {
```
Comment (freekman) [inline src/com/clouway/downloadAgent/DownloadAgent.java]: why haven you used Dependence Injection

File: src/com/clouway/downloadAgent/ProgressBar.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.clouway.downloadAgent;
+
+/**
+ * @Author Martin Milev (martinmariusmilev@gmail.com)
+ * @Date 12.07.16.
+ */
+public interface ProgressBar {
```
Comment (freekman) [inline src/com/clouway/downloadAgent/ProgressBar.java]: As i have told your colleges  in java public interfaces must have proper documentatio

**Comment by martinmilev** [comment]
Yes. 
I can't assign you, but it is ready.

File: src/com/clouway/downloadAgent/DownloadAgent.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    public void progress(int progress) {}
+  };
+
+  public long downloаdFile(String link, String path, String filename) {
+    int bitesRead = 0;
+    try {
+      URLConnection urlConnection = new URL(link).openConnection();
+      InputStream in = urlConnection.getInputStream();
+
+      OutputStream out = new FileOutputStream(path + filename + link.substring(link.lastIndexOf("."), link.length()));
```
Comment (freekman) [inline src/com/clouway/downloadAgent/DownloadAgent.java]: please extract ```path + filename + link.substring(link.lastIndexOf(".")``` in a separate variable  to enhance readability

File: test/com/clouway/downloadAgent/DownloadAgentTests.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  @Mock
+  ProgressBar progressBar;
+
+  @Test
+  public void happyPath() throws Exception {
+    DownloadAgent downloadAgent = new DownloadAgent();
+    String path = "test/com/clouway/downloadAgent/";
+    File expected = new File(path + "expected.jpg");
+    long actualSize = downloadAgent.downloаdFile(expected.toURI().toURL().toString(), path, "actual");
```
Comment (freekman) [inline test/com/clouway/downloadAgent/DownloadAgentTests.java]: again ```expected.toURI().toURL().toString()``` extract it

File: src/com/clouway/downloadAgent/DownloadAgent.java (hunk 1 above)
Comment (martinmilev) [inline src/com/clouway/downloadAgent/DownloadAgent.java]: Fixed.

File: src/com/clouway/downloadAgent/DownloadAgent.java (hunk 4 above)
Comment (martinmilev) [inline src/com/clouway/downloadAgent/DownloadAgent.java]: Fixed.

File: test/com/clouway/downloadAgent/DownloadAgentTests.java (hunk 5 above)
Comment (martinmilev) [inline test/com/clouway/downloadAgent/DownloadAgentTests.java]: Fixed.

File: src/com/clouway/downloadAgent/DownloadAgent.java (hunk 2 above)
Comment (martinmilev) [inline src/com/clouway/downloadAgent/DownloadAgent.java]: I don't actually know how to use Dependence Injection. So I did not think of it at the time. Will read more about it and try to use it in the future.

**Comment by martinmilev** [comment]
Fixes are done. Ready for review.

File: src/com/clouway/downloadAgent/DownloadAgent.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import java.io.OutputStream;
+import java.net.URL;
+import java.net.URLConnection;
+
+/**
+ * @Author Martin Milev (martinmariusmilev@gmail.com)
+ */
+public class DownloadAgent {
+  private ProgressBar progressBar;
+  private int packSize;
```
Comment (freekman) [inline src/com/clouway/downloadAgent/DownloadAgent.java]: make it final

File: src/com/clouway/downloadAgent/DownloadAgent.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+  public DownloadAgent(ProgressBar progressBar, int bites) {
```
Comment (freekman) [inline src/com/clouway/downloadAgent/DownloadAgent.java]: when you see ```int bytes ``` what do you think it is for read step size, max file size or what

File: test/com/clouway/downloadAgent/DownloadAgentTests.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * @Author Martin Milev (martinmariusmilev@gmail.com)
+ */
+public class DownloadAgentTests {
+
+  private DownloadAgent downloadAgent;
+
+  @Rule
+  public JUnitRuleMockery context = new JUnitRuleMockery();
```
Comment (freekman) [inline test/com/clouway/downloadAgent/DownloadAgentTests.java]: public field come first

File: test/com/clouway/downloadAgent/DownloadAgentTests.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+  @Mock
+  private ProgressBar progressBar;
+
+  @Before
```
Comment (freekman) [inline test/com/clouway/downloadAgent/DownloadAgentTests.java]: This is unnecessary

**Comment by freekman** [comment]
LGTM after fix comments

File: test/com/clouway/downloadAgent/DownloadAgentTests.java (hunk 9 above)
Comment (martinmilev) [inline test/com/clouway/downloadAgent/DownloadAgentTests.java]: Fixed

File: test/com/clouway/downloadAgent/DownloadAgentTests.java (hunk 8 above)
Comment (martinmilev) [inline test/com/clouway/downloadAgent/DownloadAgentTests.java]: Fixed

File: src/com/clouway/downloadAgent/DownloadAgent.java (hunk 7 above)
Comment (martinmilev) [inline src/com/clouway/downloadAgent/DownloadAgent.java]: Fixed

**Comment by freekman** [comment]
again not merged

**Comment by martinmilev** [comment]
I can't merge it. Don't have permission, or something like that.


2016-07-18 10:09 GMT+03:00 freekman <notifications@github.com>:

**Pull request merged**